# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1515616907 18000
#      Wed Jan 10 15:41:47 2018 -0500
# Node ID 466e2aab871b63601e060905049d60fd6711f905
# Parent  50545190a6049325f585926e5246f215e26e72d5
stay in debug mode on Control-C interrupts (bug #52750)

* pt-eval.cc (tree_evaluator::visit_statement): Catch
octave::interrupt_exception.  If in debug mode, recover and continue
in debug mode.  Otherwise, rethrow exception.

diff --git a/libinterp/parse-tree/pt-eval.cc b/libinterp/parse-tree/pt-eval.cc
--- a/libinterp/parse-tree/pt-eval.cc
+++ b/libinterp/parse-tree/pt-eval.cc
@@ -2324,16 +2324,26 @@ namespace octave
             // FIXME: We want to use error_with_id here so that give users
             // control over this error message but error_with_id will
             // require some memory allocations.  Is there anything we can
             // do to make those more likely to succeed?
 
             error_with_id ("Octave:bad-alloc",
                            "out of memory or dimension too large for Octave's index type");
           }
+        catch (const octave::interrupt_exception&)
+          {
+            // If we are debugging, then continue with next statement.
+            // Otherwise, jump out of here.
+
+            if (debug_mode)
+              interpreter::recover_from_exception ();
+            else
+              throw;
+          }
       }
   }
 
   void
   tree_evaluator::visit_statement_list (tree_statement_list& lst)
   {
     // FIXME: commented out along with else clause below.
     // static octave_value_list empty_list;
